File: r/svm.R
# ============================================
# 🛡️ Support Vector Classification (SVC)
# ============================================

# 📂 1. Load Dataset
dataset <- read.csv('../data/Social_network_ads.csv')
# Columns: Age, EstimatedSalary, Purchased

# 🔀 2. Split into Training and Test Sets
library(caTools)
set.seed(123)
split <- sample.split(dataset$Purchased, SplitRatio = 0.8)
train_set <- subset(dataset, split == TRUE)
test_set  <- subset(dataset, split == FALSE)

# ⚖️ 3. Feature Scaling
train_set[, 1:2] <- scale(train_set[, 1:2])
test_set[,  1:2] <- scale(test_set[,  1:2])

# 🧠 4. Train SVM Classifier
library(e1071)
classifier <- svm(
  x    = train_set[, 1:2],
  y    = train_set$Purchased,
  type = 'C-classification',
  kernel = 'radial'  # default; change to 'linear' or others if desired
)

# 🔍 5. Predict on Test Set
y_pred <- predict(classifier, newdata = test_set[, 1:2])

# 📊 6. Confusion Matrix
cm <- table(Actual = test_set$Purchased, Predicted = y_pred)
print(cm)

# 📈 7. Decision Boundary Plot Function
plot_svc_boundary <- function(data, title) {
  # Grid for plotting
  X1 <- seq(min(data$Age) - 1, max(data$Age) + 1, by = 0.01)
  X2 <- seq(min(data$EstimatedSalary) - 1, max(data$EstimatedSalary) + 1, by = 0.01)
  grid <- expand.grid(Age = X1, EstimatedSalary = X2)
  
  # Predict grid labels
  grid$Purchased <- predict(classifier, newdata = grid)
  
  # Base plot
  plot(
    data$Age, data$EstimatedSalary,
    type = 'n',
    main = title,
    xlab = 'Age (scaled)',
    ylab = 'Estimated Salary (scaled)',
    xlim = c(min(X1), max(X1)),
    ylim = c(min(X2), max(X2))
  )
  
  # Add decision boundary
  contour(
    X1, X2,
    matrix(as.numeric(grid$Purchased), length(X1), length(X2)),
    add = TRUE
  )
  
  # Plot grid points
  points(
    grid$Age, grid$EstimatedSalary,
    pch = '.', 
    col = ifelse(grid$Purchased == 1, 'springgreen3', 'tomato'),
    cex = 0.5
  )
  
  # Plot actual observations
  points(
    data$Age, data$EstimatedSalary,
    pch = 21,
    bg  = ifelse(data$Purchased == 1, 'green4', 'red3'),
    col = 'black',
    cex = 1
  )
}

# 🖼️ 8. Visualize Training Set
plot_svc_boundary(train_set, title = 'SVC (Training Set)')

# 🖼️ 9. Visualize Test Set
plot_svc_boundary(test_set, title = 'SVC (Test Set)')

File: python/svm.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jun 26 13:22:00 2025
Author: Chukwuka Chijioke Jerry

Support Vector Classification on Social Network Ads
"""

# ===========================
# 📦 Import Libraries
# ===========================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from matplotlib.colors import ListedColormap
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, accuracy_score

# ===========================
# 📂 Load Dataset
# ===========================
dataset = pd.read_csv('../data/social_network_ads.csv')
X = dataset.iloc[:, :2].values  # Age, EstimatedSalary
y = dataset.iloc[:, -1].values  # Purchased

# ===========================
# 🔀 Split into Train/Test
# ===========================
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0
)

# ===========================
# ⚖️ Feature Scaling
# ===========================
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# ===========================
# 🧠 Train SVC Model
# ===========================
classifier = SVC(kernel='rbf', random_state=0)
classifier.fit(X_train, y_train)

# ===========================
# 🔍 Predict & Evaluate
# ===========================
y_pred = classifier.predict(X_test)

# Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
sns.heatmap(cm, annot=True, fmt='d', cmap='viridis', cbar=False)
plt.title('Confusion Matrix')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.tight_layout()
plt.show()

# Accuracy
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")

# ===========================
# 📊 Decision Boundary Plotter
# ===========================
def plot_decision_boundary(X_set, y_set, title):
    X1, X2 = np.meshgrid(
        np.arange(X_set[:, 0].min() - 1, X_set[:, 0].max() + 1, 0.01),
        np.arange(X_set[:, 1].min() - 1, X_set[:, 1].max() + 1, 0.01)
    )
    Z = classifier.predict(np.c_[X1.ravel(), X2.ravel()]).reshape(X1.shape)

    plt.contourf(
        X1, X2, Z,
        alpha=0.3,
        cmap=ListedColormap(('red', 'green'))
    )
    plt.xlim(X1.min(), X1.max())
    plt.ylim(X2.min(), X2.max())

    for idx, label in enumerate(np.unique(y_set)):
        plt.scatter(
            X_set[y_set == label, 0],
            X_set[y_set == label, 1],
            c=('red', 'green')[idx],
            label=f'Class {label}'
        )
    plt.title(title)
    plt.xlabel('Age (scaled)')
    plt.ylabel('Estimated Salary (scaled)')
    plt.legend()
   
    plt.tight_layout()
    plt.show()

# ===========================
# 🖼️ Plot for Training & Test Sets
# ===========================
plot_decision_boundary(X_train, y_train, 'SVC (Training Set)')
plot_decision_boundary(X_test, y_test, 'SVC (Test Set)')
